### PR TITLE
feat!: fx and interscheme implementation

### PR DESCRIPTION
**Pull request by kleyow**
**Changes:**
- Currency Conversion (FX)
  - Database changes (knex migrations) for FX
  - Updated prepare and position handlers to support FX kafka messages
  - Updated timeout handler to timeout dependent transfers incase of FX
  - Implemented batch mode for the timeout and abort position messages along with FX actions
- Interscheme
  - Proxy cache implementation using redis
  - Parameterize switch id
  - Extend admin api to support proxy participants
  - Added new transfer internal state (RESERVED_FORWARDED) to support interscheme
  - Adjusted position handler implementation for proxied participants
  - Adjusted validations in prepare handler to support proxied participants
- Performance optimisations

### CODE REVIEW

File: src/shared/fspiopErrorFactory.js [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+      ERROR_MESSAGES.fxTransferNonReservedState,
+      cause, replyTo
+    )
+  },
+
+  fxTransferExpired: (cause = null, replyTo = '') => {
+    return Factory.createFSPIOPError(
+      Enums.FSPIOPErrorCodes.TRANSFER_EXPIRED,
+      ERROR_MESSAGES.fxTransferExpired,
+      cause = null, replyTo = ''
```
Comment (shashi165) [inline src/shared/fspiopErrorFactory.js:76]:
```suggestion
      cause, replyTo
```

File: .circleci/config.yml [hunk 2]
```diff
@@ -1,1059 +1,11 @@
...
-            - test-lint
-            - test-unit
-            - test-coverage
-            - test-integration
-            - test-functional
-            - vulnerability-check
-            - audit-licenses
-            - license-scan
-            - image-scan
+      - build/workflow:
```
Comment (elnyry-sam-k) [inline .circleci/config.yml:8]: Like this change but need to make sure we're not missing any step. please confirm. I didn't find the license scan and another scan there on the orb, could've missed it but would like to confirm

File: documentation/fx-implementation/README.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
+# FX Implementation
+
+## Proof of Concept (PoC) Implementation for Payer-Side Currency Conversion (Happy Path Only)
```
Comment (elnyry-sam-k) [inline documentation/fx-implementation/README.md:3]: Does this still hold?

File: package.json [hunk 4]
```diff
@@ -79,25 +81,26 @@
...
     "@hapi/good": "9.0.1",
     "@hapi/hapi": "21.3.10",
-    "@hapi/basic": "7.0.2",
     "@hapi/inert": "7.1.0",
     "@hapi/joi": "17.1.1",
     "@hapi/vision": "7.0.3",
-    "@hapi/catbox-memory": "6.0.2",
-    "@mojaloop/database-lib": "11.0.5",
     "@mojaloop/central-services-error-handling": "13.0.1",
     "@mojaloop/central-services-health": "15.0.0",
```
Comment (shashi165) [inline package.json:92]: is it missing dependency @mojaloop/sdk-standard-components required by @mojaloop/central-services-error-handling

**Comment by shashi165** [comment]
Need to fix the sonarcloud issues and run the tests. integration tests are failing locally

File: .circleci/config.yml (hunk 2 above)
Comment (kalinkrustev) [reply .circleci/config.yml:8]: License and vulnerability scans is present. See example here https://app.circleci.com/pipelines/github/mojaloop/central-ledger/2809/workflows/b9859b2a-253d-483b-80b8-3e395393bd11

Comment (elnyry-sam-k) [reply .circleci/config.yml:8]: Thanks!

File: documentation/fx-implementation/README.md (hunk 3 above)
Comment (kleyow) [reply documentation/fx-implementation/README.md:3]: Hey Sam. Yes currently only payer side conversion is fully supported.

Comment (elnyry-sam-k) [reply documentation/fx-implementation/README.md:3]: Thanks @kleyow , also Is it still a PoC implementation?

Comment (kleyow) [reply documentation/fx-implementation/README.md:3]: @elnyry-sam-k Personally, I'm not sure.

File: documentation/state-diagrams/transfer-internal-states.plantuml [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  state INVALID {
+  }
+}
+
+RECEIVED_FULFIL_DEPENDENT : only FX-transfer
+RECEIVED_FULFIL : only transfer
+
+[*] --> RECEIVED_PREPARE : Transfer Prepare Request [Prepare handler] \n <i>(validation & dupl.check passed)</i>
+[*] --> INVALID : Validation failed \n [Prepare handler]
+RECEIVED_PREPARE --> RESERVED : [Position handler]: Liquidity check passed, \n funds reserved
```
Comment (elnyry-sam-k) [inline documentation/state-diagrams/transfer-internal-states.plantuml:48]: how can a Payee FSP respond to something thats not reserved at the switch and forwarded to it (the Payee FSP)? As in, if the Payee is able to reject it, it should've received it from the switch, where it should've been reserved. Isn't it?

File: migrations/600700_fxTransferFulfilment.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+ should be listed with a '*' in the first column. People who have
+ contributed from an organization can be listed under the organization
+ that actually holds the copyright for their contributions (see the
+ Gates Foundation organization for an example). Those individuals should have
+ their names indented and be marked with a '-'. Email address can be added
+ optionally within square brackets <email>.
+ * Gates Foundation
+ - Name Surname <name.surname@gatesfoundation.com>
+
+ * ModusBox
```
Comment (elnyry-sam-k) [inline migrations/600700_fxTransferFulfilment.js:21]: Infitx? same in other places

File: documentation/state-diagrams/transfer-internal-states.plantuml [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+RESERVED --> RECEIVED_REJECT : Reject callback from Payee with status "ABORTED"
+RECEIVED_PREPARE --> RECEIVED_ERROR : Transfer Error callback from Payee
```
Comment (elnyry-sam-k) [inline documentation/state-diagrams/transfer-internal-states.plantuml:50]: how can a Payee FSP respond to something thats not reserved at the switch and forwarded to it (the Payee FSP)? As in, if the Payee is able to reject it, it should've received it from the switch, where it should've been reserved. Isn't it?

File: src/domain/fx/index.js [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    const result = await FxTransferModel.fxTransfer.updateFxPrepareReservedForwarded(commitRequestId)
+    histTimerTransferServicePrepareEnd({ success: true, funcName: 'forwardedFxPrepare' })
+    return result
+  } catch (err) {
+    histTimerTransferServicePrepareEnd({ success: false, funcName: 'forwardedFxPrepare' })
+    throw ErrorHandler.Factory.reformatFSPIOPError(err)
+  }
+}
+
+// TODO: Need to implement this for fxTransferError
```
Comment (elnyry-sam-k) [inline src/domain/fx/index.js:73]: Is this implementation still needed?

File: src/domain/fx/index.js [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+const handleFulfilResponse = async (transferId, payload, action, fspiopError) => {
+  const histTimerTransferServiceHandlePayeeResponseEnd = Metrics.getHistogram(
+    'fx_domain_transfer',
+    'prepare - Metrics for fx transfer domain',
+    ['success', 'funcName']
+  ).startTimer()
+
+  try {
+    await FxTransferModel.fxTransfer.saveFxFulfilResponse(transferId, payload, action, fspiopError)
+    // TODO: Need to return a result if we need
```
Comment (elnyry-sam-k) [inline src/domain/fx/index.js:46]: Is this completed or not needed anymore?

File: src/domain/fx/cyril.js [hunk 10]
```diff
@@ -0,0 +1,466 @@
...
+            isFxTransferStateChange: true,
+            commitRequestId: receivingFxpRecord.commitRequestId,
+            participantCurrencyId: proxyParticipantAccountDetails.participantCurrencyId,
+            amount: -receivingFxpRecord.sourceAmount
+          })
+        }
+      }
+    }
+
+    // TODO: Remove entries from watchlist
```
Comment (elnyry-sam-k) [inline src/domain/fx/cyril.js:448]: still pending?

Comment (kleyow) [reply src/domain/fx/cyril.js:448]: @vijayg10 Can this be removed/moved to a story. Not sure of context.

File: src/domain/fx/index.js (hunk 9 above)
Comment (kleyow) [reply src/domain/fx/index.js:46]: @vijayg10 Can this be removed/moved to a story? Not sure of context.
Going off of git blame for the comment.

File: src/domain/fx/index.js (hunk 8 above)
Comment (kleyow) [reply src/domain/fx/index.js:73]: @vijayg10 Can this be removed/moved to a story? Not sure of context.
Going off of git blame for the comment.

File: migrations/600700_fxTransferFulfilment.js (hunk 6 above)
Comment (kleyow) [reply migrations/600700_fxTransferFulfilment.js:21]: Will be addressed in later PR due to diff size. 
https://github.com/mojaloop/central-ledger/pull/1138

File: documentation/state-diagrams/transfer-internal-states.plantuml (hunk 7 above)
Comment (kleyow) [reply documentation/state-diagrams/transfer-internal-states.plantuml:50]: :thinking: :thinking: :thinking: 

I believe if a fxTransfer is reserved and a subsequent transfer is prepared and stuck in `RECEIVED_PREPARE` (example transfer position handler hangs). If the FXP returns an `/error` the handlers will abort both the fxTransfer and transfer thus performing the state change for the transfer `RECEIVED_PREPARE --> RECEIVED_ERROR`.

I believe that is the hypothetical case :thinking: 

@geka-evk 
@vijayg10 

Anything to add or is this a mistake.

Comment (elnyry-sam-k) [reply documentation/state-diagrams/transfer-internal-states.plantuml:50]: Thanks Kevin. Even in that case, saying "Reject callback from Payee with status "ABORTED" doesn't seem apt; maybe a concise version of what you said above..

Comment (kleyow) [reply documentation/state-diagrams/transfer-internal-states.plantuml]: Talked it out with Vijay. This is just a mistake. Removed.

File: src/domain/fx/index.js (hunk 8 above)
Comment (kleyow) [reply src/domain/fx/index.js]: Created a follow up story. Removed.

File: src/domain/fx/index.js (hunk 9 above)
Comment (kleyow) [reply src/domain/fx/index.js]: Removed.

File: src/domain/fx/cyril.js (hunk 10 above)
Comment (kleyow) [reply src/domain/fx/cyril.js]: Removed.

File: documentation/state-diagrams/transfer-internal-states.plantuml (hunk 7 above)
Comment (elnyry-sam-k) [reply documentation/state-diagrams/transfer-internal-states.plantuml]: Thanks @kleyow !

**Review by elnyry-sam-k — APPROVED** [review]

**Comment by elnyry-sam-k** [comment]
Merging as requested by author

**Pull request merged**